src/UKHO.ADDS.EFS.Builder.S100/Pipelines/Assemble/DownloadFilesNode.cs

```diff
@@ -1,10 +1,117 @@
-﻿using UKHO.ADDS.Infrastructure.Pipelines;
+﻿using UKHO.ADDS.Clients.FileShareService.ReadOnly;
+using UKHO.ADDS.Clients.FileShareService.ReadOnly.Models;
+using UKHO.ADDS.EFS.Builder.S100.Pipelines.Assemble.Logging;
+using UKHO.ADDS.Infrastructure.Pipelines;
 using UKHO.ADDS.Infrastructure.Pipelines.Nodes;
+using UKHO.ADDS.Infrastructure.Results;
 
 namespace UKHO.ADDS.EFS.Builder.S100.Pipelines.Assemble
 {
     internal class DownloadFilesNode : ExchangeSetPipelineNode
     {
-        protected override Task<NodeResultStatus> PerformExecuteAsync(IExecutionContext<ExchangeSetPipelineContext> context) => Task.FromResult(NodeResultStatus.NotRun);
+        private readonly IFileShareReadOnlyClient _fileShareReadOnlyClient;
+        private ILogger _logger;
+
+        private const long FileSizeInBytes = 10485760;
+        private const string ProductName = "ProductName";
+        private const string EditionNumber = "EditionNumber";
+        private const string UpdateNumber = "UpdateNumber";
+
+        public DownloadFilesNode(IFileShareReadOnlyClient fileShareReadOnlyClient)
+        {
+            _fileShareReadOnlyClient = fileShareReadOnlyClient ?? throw new ArgumentNullException(nameof(fileShareReadOnlyClient));
+        }
+
+        protected override async Task<NodeResultStatus> PerformExecuteAsync(IExecutionContext<ExchangeSetPipelineContext> context)
+        {
+            try
+            {
+                var batchDetails = context.Subject.BatchDetails;
+                var downloadPath = Path.Combine(context.Subject.WorkSpaceRootPath, "fssdata");
+
+                _logger = context.Subject.LoggerFactory.CreateLogger<DownloadFilesNode>();
+
+                EnsureDownloadDirectoryExists(downloadPath);
+
+                var result = await DownloadLatestBatchFilesAsync(context, SelectLatestBatchesByProductEditionAndUpdate(batchDetails));
+
+                return result;
+            }
+            catch (Exception ex)
+            {
+                _logger.LogDownloadFilesNodeFailed(ex.Message);
+
+                return NodeResultStatus.Failed;
+            }
+        }
+
+        private static IEnumerable<BatchDetails> SelectLatestBatchesByProductEditionAndUpdate(IEnumerable<BatchDetails> batchDetails)
+        {
+            var latestBatches = batchDetails
+                .Where(b => b.Attributes != null)
+                .GroupBy(b =>
+                    new
+                    {
+                        ProductName = b.Attributes.FirstOrDefault(a => a.Key == ProductName)?.Value,
+                        EditionNumber = b.Attributes.FirstOrDefault(a => a.Key == EditionNumber)?.Value,
+                        UpdateNumber = b.Attributes.FirstOrDefault(a => a.Key == UpdateNumber)?.Value
+                    })
+                .Select(g => g.OrderByDescending(b => b.BatchPublishedDate).First())
+                .ToList();
+
+            return latestBatches;
+        }
```

Comment:
try below change, we can further optimize it, consider extracting attribute values once per batch to avoid repeated lookups, and use a tuple for grouping to improve readability and performance. Additionally, since the method already returns an IEnumerable, you can remove the unnecessary ToList() call to allow for deferred execution and potentially better performance.

```suggestion
        private static IEnumerable<BatchDetails> SelectLatestBatchesByProductEditionAndUpdate(IEnumerable<BatchDetails> batchDetails)
{
    return batchDetails
        .Where(b => b.Attributes != null)
        .Select(b =>
        {
            var productName = b.Attributes.FirstOrDefault(a => a.Key == ProductName)?.Value;
            var editionNumber = b.Attributes.FirstOrDefault(a => a.Key == EditionNumber)?.Value;
            var updateNumber = b.Attributes.FirstOrDefault(a => a.Key == UpdateNumber)?.Value;
            return new { Batch = b, productName, editionNumber, updateNumber };
        })
        .GroupBy(x => (x.productName, x.editionNumber, x.updateNumber))
        .Select(g => g.OrderByDescending(x => x.Batch.BatchPublishedDate).First().Batch);
}
```

Reply:
Fixed review comments